eZ/Publish/Core/MVC/Symfony/Controller/Content/QueryController.php

```diff
@@ -81,4 +87,55 @@ private function runQuery(ContentView $view, $method)
         );
         $view->addParameters([$view->getParameter('query')['assign_results_to'] => $searchResults]);
     }
+
+    /**
+     * @param \eZ\Publish\Core\MVC\Symfony\View\ContentView $view
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @return \eZ\Publish\Core\MVC\Symfony\View\ContentView
+     */
+    public function pagingQueryAction(ContentView $view, Request $request)
+    {
+        $this->runPagingQuery($view, $request);
+
+        return $view;
+    }
+
+    /**
+     * @param \eZ\Publish\Core\MVC\Symfony\View\ContentView $view
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     */
+    private function runPagingQuery(ContentView $view, Request $request)
+    {
+        $pager = new Pagerfanta(
+            $this->getAdapter($this->contentViewQueryTypeMapper->map($view))
+        );
+
+        $queryParameters = $view->getParameter('query');
+        $pager->setMaxPerPage($queryParameters['limit']);
+
+        $page_param = 'page';
+        if (array_key_exists('page_param', $queryParameters)) {
+            $page_param = $queryParameters['page_param'];
+        }
+
+        $page = $request->get($page_param, 1);
+        if ($page > 1) {
+            $pager->setCurrentPage($page);
+        }
+
+        $view->addParameters([$queryParameters['assign_results_to'] => $pager]);
+    }
+
+    /**
+     * @param Query|LocationQuery $query
+     * @return ContentSearchHitAdapter|LocationSearchHitAdapter
+     */
+    private function getAdapter(Query $query)
```

Comment:
Here you should specify strictly typed return type, however you have multiple types returned. For this case the proper solution is to return `\Pagerfanta\Adapter\AdapterInterface` (side: multiple return types suggest a need to redesign the whole thing, but let's not over-engineer it now).

So in this line there should be `: AdapterInterface` with a proper import at the top of the file. PhpDoc however needs to have FQCN as usual (applies also to `@param` above).